File: src/main/java/hu/gerviba/webschop/model/OpeningEntity.java
package hu.gerviba.webschop.model;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.OrderBy;
import javax.persistence.Table;
import javax.validation.constraints.Min;
import javax.validation.constraints.Size;

import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.JsonIdentityReference;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;

import hu.gerviba.webschop.service.OpeningService;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Entity
@NoArgsConstructor
@Table(name = "openings")
@SuppressWarnings("serial")
public class OpeningEntity implements Serializable {
    
    @Id
    @Column
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;
    
    @Column
    private int timeIntervals;
    
    @Column
    private Long dateStart;
    
    @Column
    private Long dateEnd;

    @Column
    private Long orderStart;
    
    @Column
    private Long orderEnd;

    @Column
    private String prUrl;
    
    @Column
    private String eventDescription;

    @Column(length = 255)
    @Size(max = 255)
    private String feeling;
    
    @JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
    @JsonIdentityReference(alwaysAsId = true)
    @ManyToOne(fetch = FetchType.LAZY)
    private CircleEntity circle;

    @Column
    @OrderBy("id ASC")
    @OneToMany(mappedBy = "opening", fetch = FetchType.LAZY, orphanRemoval = true)
    private List<TimeWindowEntity> timeWindows;
    
    @Column
    @Min(0)
    private int maxOrder;

    @Column
    @Min(0)
    private int maxOrderPerInterval;

    @Column
    @Min(0)
    private int maxExtraPerInterval;
    
    @Column
    @Min(0)
    private int intervalLength;
    
    @Column
    @Min(0)
    private int orderCount;

    @Column(nullable = false)
    private Integer maxAlpha = 0;

    @Column(nullable = false)
    private Integer maxBeta = 0;

    @Column(nullable = false)
    private Integer maxGamma = 0;

    @Column(nullable = false)
    private Integer maxDelta = 0;

    @Column(nullable = false)
    private Integer maxLambda = 0;

    @Column(nullable = false)
    private Integer usedAlpha = 0;

    @Column(nullable = false)
    private Integer usedBeta = 0;

    @Column(nullable = false)
    private Integer usedGamma = 0;

    @Column(nullable = false)
    private Integer usedDelta = 0;

    @Column(nullable = false)
    private Integer usedLambda = 0;

    private static final long MILLIS_TO_MINS = 60000;
    
    /**
     * This method is for testing purposes
     */
    public OpeningEntity(Long dateStart, Long dateEnd, Long orderStart, Long orderEnd, String prUrl, String eventDescription,
            String feeling, CircleEntity circle, int maxOrder, int maxOrderPerInterval, int maxExtraPerInterval,
            int intervalLength) {
        this.dateStart = dateStart;
        this.dateEnd = dateEnd;
        this.orderStart = orderStart;
        this.orderEnd = orderEnd;
        this.prUrl = prUrl;
        this.eventDescription = eventDescription;
        this.feeling = feeling;
        this.circle = circle;
        this.maxOrder = maxOrder;
        this.maxOrderPerInterval = maxOrderPerInterval;
        this.maxExtraPerInterval = maxExtraPerInterval;
        this.intervalLength = intervalLength;
        this.orderCount = 0;
    }

    public void generateTimeWindows(OpeningService openings) {
        timeWindows = new ArrayList<>();
        if (this.getIntervalLength() <= 0) {
            appendTimeWindow(openings, this.getDateStart());
            return;
        }
        
        for (long time = this.getDateStart(); time < this.getDateEnd(); time += this.getIntervalLength() * MILLIS_TO_MINS)
            appendTimeWindow(openings, time);
    }

    private void appendTimeWindow(OpeningService openings, long time) {
        TimeWindowEntity tw = new TimeWindowEntity(null, this, 
                OpeningService.DATE_FORMATTER_HH_MM.format(time)
                    + " - " + OpeningService.DATE_FORMATTER_HH_MM.format(time + this.getIntervalLength() * MILLIS_TO_MINS),
                time,
                this.getMaxOrderPerInterval(),
                this.getMaxExtraPerInterval());
        timeWindows.add(tw);
        openings.saveTimeWindow(tw);
    }

    public boolean isInInterval(long timeMillis) {
        return orderStart <= timeMillis && orderEnd >= timeMillis;
    }
    
}

File: src/main/kotlin/hu/kirdev/schpincer/model/OpeningEntity.kt
package hu.kirdev.schpincer.model

import com.fasterxml.jackson.annotation.JsonBackReference
import com.fasterxml.jackson.annotation.JsonIdentityInfo
import com.fasterxml.jackson.annotation.JsonIdentityReference
import com.fasterxml.jackson.annotation.ObjectIdGenerators.PropertyGenerator
import hu.kirdev.schpincer.service.OpeningService
import java.io.Serializable
import jakarta.persistence.*
import jakarta.validation.constraints.Size

const val MILLIS_TO_MINS: Long = 60000

@Entity
@Table(name = "openings")
data class OpeningEntity(

        @Id
        @Column
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        var id: Long = 0,

        @Column
        var timeIntervals: Int = 0,

        @Column
        var dateStart: Long,

        @Column
        var dateEnd: Long,

        @Column
        var orderStart: Long,

        @Column
        var orderEnd: Long,

        @Column
        var prUrl: String? = null,

        @Column
        var eventDescription: String? = null,

        @Column(length = 255)
        var feeling: @Size(max = 255) String? = null,

        @field:JsonIdentityInfo(generator = PropertyGenerator::class, property = "id")
        @field:JsonIdentityReference(alwaysAsId = true)
        @field:ManyToOne(fetch = FetchType.LAZY)
        @get:JsonIdentityInfo(generator = PropertyGenerator::class, property = "id")
        @get:JsonIdentityReference(alwaysAsId = true)
        @get:ManyToOne(fetch = FetchType.LAZY)
        @JsonBackReference
        var circle: CircleEntity? = null,

        @Column
        @OrderBy("id ASC")
        @OneToMany(mappedBy = "opening", fetch = FetchType.LAZY, orphanRemoval = true)
        var timeWindows: MutableList<TimeWindowEntity> = mutableListOf(),

        @Column
        var maxOrder: Int = 0,

        @Column
        var maxOrderPerInterval: Int = 0,

        @Column
        var maxExtraPerInterval: Int = 0,

        @Column
        var intervalLength: Int = 0,

        @Column
        var orderCount: Int = 0,

        @Column(nullable = false)
        var maxAlpha: Int = 0,

        @Column(nullable = false)
        var maxBeta: Int = 0,

        @Column(nullable = false)
        var maxGamma: Int = 0,

        @Column(nullable = false)
        var maxDelta: Int = 0,

        @Column(nullable = false)
        var maxLambda: Int = 0,

        @Column(nullable = false)
        var usedAlpha: Int = 0,

        @Column(nullable = false)
        var usedBeta: Int = 0,

        @Column(nullable = false)
        var usedGamma: Int = 0,

        @Column(nullable = false)
        var usedDelta: Int = 0,

        @Column(nullable = false)
        var usedLambda: Int = 0,

        @Column(nullable = false, columnDefinition = "bigint(20) DEFAULT 0")
        var compensationTime: Long = 0

) : Serializable {

    fun generateTimeWindows(openings: OpeningService) {
        if (this.intervalLength <= 0) {
            appendTimeWindow(openings, this.dateStart)
            return
        }
        var time: Long = this.dateStart
        while (time < this.dateEnd) {
            appendTimeWindow(openings, time)
            time += this.intervalLength * MILLIS_TO_MINS
        }

        val tw = TimeWindowEntity(0, this, "Túlóra", this.dateEnd, 0, Integer.MIN_VALUE)
        timeWindows.add(tw)
        openings.saveTimeWindow(tw)
    }

    private fun appendTimeWindow(openings: OpeningService, time: Long) {
        val tw = TimeWindowEntity(0, this,
                OpeningService.DATE_FORMATTER_HH_MM.format(time)
                        + " - " + OpeningService.DATE_FORMATTER_HH_MM.format(time + this.intervalLength * MILLIS_TO_MINS),
                time,
                this.maxOrderPerInterval,
                this.maxExtraPerInterval)
        timeWindows.add(tw)
        openings.saveTimeWindow(tw)
    }

    fun isInInterval(timeMillis: Long) = timeMillis in orderStart..orderEnd

}
